### PR TITLE
EVG-7200: Filter favorites out of all projects

### PR DESCRIPTION
**Pull request by tgrander**
Having favorited projects also in all projects caused duplicate key issues when rendering the data in the projects dropdown. This PR filters favorite projects out of all projects.

### CODE REVIEW

**Comment by SupaJoon** [comment]
The code in the resolver looks right but the test is failing

**Comment by tgrander** [comment]
> The code in the resolver looks right but the test is failing

yeah i was trying to seed the user data in the `integration_spec.json` file but realized that was the wrong user. The user these tests act on is seeded in `integration_test.go` in `SetupSuite`

**Comment by tgrander** [comment]
> It seems wrong to call something Projects.all but not return all projects there. This filtering should happen in the client

i don't agree that the filtering should happen in the client. If we know the shape of the data we want the UI to consume then gql should deliver the data in that shape. Regarding the name of what the field is called, do you have another suggestion?

**Comment by john-m-liu** [comment]
I suppose you can call it something along the lines of other_projects. However what you're doing at the moment is mapping over 2 separate arrays. Would it work to have this query return all projects with favorites sorted to the top, and let the UI just map over the single array?

File: graphql/resolvers.go [hunk 1]
```diff
@@ -115,16 +115,15 @@ func (r *queryResolver) Projects(ctx context.Context) (*Projects, error) {
 			Owner:       p.Owner,
 		}
 
+		if isFavorite := util.StringSliceContains(favoriteIds, p.Identifier); isFavorite {
```
Comment (tychoish) [inline graphql/resolvers.go]: you don't need to declare isFavorite here. 

I also wouldn't declare favoriteIds as its own variable, and just reference usr.FavoriteProjects. Extra variables impede readability.

**Comment by tgrander** [comment]
> I suppose you can call it something along the lines of other_projects. However what you're doing at the moment is mapping over 2 separate arrays. Would it work to have this query return all projects with favorites sorted to the top, and let the UI just map over the single array?

When it comes to how the data is consumed by the UI, which is consuming a favorites array and an other_projects array, it makes perfect sense to return two separate arrays. The only case that returning a single array and filtering on the client makes sense is that when a new favorite is added, it will automatically be filtered out of `other_projects` array. The two arrays method, on the other hand, requires removing the project from `other_projects` after the mutation has succeeded. If two arrays were returned then the projects query would return an array of strings representing favorites, which would then be used to filter favorites.

File: graphql/resolvers.go (hunk 1 above)
Comment (tgrander) [reply graphql/resolvers.go]: `isFavorite` does not _need_ to be declared here but it also does not hurt and adds semantic value imo

**Comment by john-m-liu** [comment]
That sounds to me like a stronger argument for returning a single array, but this seems fine to me as long as you rename the field

File: graphql/resolvers.go (hunk 1 above)
Comment (tychoish) [reply graphql/resolvers.go]: I'm pretty convinced that the determent to readability of declaring extra variables is not worth any potential benefit of added semantics/annotations.

Comment (tgrander) [reply graphql/resolvers.go]: i think `isFavorite` increases readability because it reads like english. `if isFavorite .... { do this }`

Comment (bsamek) [reply graphql/resolvers.go]: As a similar example, PEP8 disagrees. They argue that `if greeting is True` is worse than `if greeting == true` or `if greeting`, even though the first is the most English. The `if isFavorite` requires reading to the end of the line, and is definitely not idiomatic go. I had to read the line a couple times before I got it. I'm used to reading `if <expression>` in many programming languages, not `if <thing> := <expression>; <thing>`, which is more things for my eyes to parse.

The usual way to use this idiom in go is:
```
if err := <function that returns error>; err != nil {
// handle error
}
```

And somewhat less usually, something like:
```
if val, ok := <thing with two return values>; ok {
// handle val, which only exists with ok is true
}
```

It's meant for creating a branch where one does not already exist. In this case though there's an expression that already evaluates to a boolean. The advantage here is that the code is somewhat more self-documenting, but I find it hard to read, and it is probably better in a comment.

**Comment by tgrander** [comment]
> That sounds to me like a stronger argument for returning a single array, but this seems fine to me as long as you rename the field

i will go single array route because i like idea of favorites auto updating when the UI re-renders

**Comment by tgrander** [comment]
> > That sounds to me like a stronger argument for returning a single array, but this seems fine to me as long as you rename the field
> 
> i will go single array route because i like idea of favorites auto updating when the UI re-renders

Wait, @john-m-liu, I just realized that returning one array of all projects won't work because of the shape of the grouped projects, which is not an array of Project types. It's nested objects. This is why i went this route in the first place. on the client we would have to use a nested loop to check each project in grouped projects. This shape i strongly believe should remain the same.

File: graphql/resolvers.go (hunk 1 above)
Comment (tgrander) [reply graphql/resolvers.go]:
```
if itMattersThisMuch := GetStrongOpinions(); itMattersThisMuch {
  then by all means I shall change it :)
}
```

**Pull request merged**